tsdb/series_partition.go

```diff
@@ -0,0 +1,665 @@
+			continue
+		}
+		ids[i] = id
+	}
+	p.mu.RUnlock()
+
+	// Exit if all series for this partition already exist.
+	if !writeRequired {
+		return nil
+	}
+
+	type keyRange struct {
+		id     uint64
+		offset int64
+	}
+	newKeyRanges := make([]keyRange, 0, len(keys))
+
+	// Obtain write lock to create new series.
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if p.closed {
+		return ErrSeriesPartitionClosed
+	}
+
+	// Track offsets of duplicate series.
+	newIDs := make(map[string]uint64, len(ids))
+
+	for i := range keys {
+		// Skip series that don't belong to the partition or have already been created.
+		if keyPartitionIDs[i] != p.id || ids[i] != 0 {
+			continue
+		}
+
+		// Re-attempt lookup under write lock.
+		key := keys[i]
+		if ids[i] = newIDs[string(key)]; ids[i] != 0 {
+			continue
+		} else if ids[i] = p.index.FindIDBySeriesKey(p.segments, key); ids[i] != 0 {
+			continue
+		}
+
+		// Write to series log and save offset.
+		id, offset, err := p.insert(key)
+		if err != nil {
+			return err
+		}
+
+		// Append new key to be added to hash map after flush.
+		ids[i] = id
+		newIDs[string(key)] = id
+		newKeyRanges = append(newKeyRanges, keyRange{id, offset})
+	}
+
+	// Flush active segment writes so we can access data in mmap.
+	if segment := p.activeSegment(); segment != nil {
+		if err := segment.Flush(); err != nil {
+			return err
+		}
+	}
+
+	// Add keys to hash map(s).
+	for _, keyRange := range newKeyRanges {
+		p.index.Insert(p.seriesKeyByOffset(keyRange.offset), keyRange.id, keyRange.offset)
+	}
+
+	// Check if we've crossed the compaction threshold.
+	if !p.compacting && p.CompactThreshold != 0 && p.index.InMemCount() >= uint64(p.CompactThreshold) {
+		p.compacting = true
+		logger := p.Logger.With(zap.String("path", p.path))
+		logger.Info("beginning series partition compaction")
+
+		startTime := time.Now()
+		p.wg.Add(1)
+		go func() {
+			defer p.wg.Done()
+
+			if err := NewSeriesPartitionCompactor().Compact(p); err != nil {
+				logger.With(zap.Error(err)).Error("series partition compaction failed")
+			}
+
+			logger.With(zap.Duration("elapsed", time.Since(startTime))).Info("completed series partition compaction")
+
+			// Clear compaction flag.
+			p.mu.Lock()
+			p.compacting = false
+			p.mu.Unlock()
+		}()
+	}
+
+	return nil
+}
+
+// DeleteSeriesID flags a series as permanently deleted.
+// If the series is reintroduced later then it must create a new id.
+func (p *SeriesPartition) DeleteSeriesID(id uint64) error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if p.closed {
+		return ErrSeriesPartitionClosed
+	}
+
+	// Already tombstoned, ignore.
+	if p.index.IsDeleted(id) {
+		return nil
+	}
+
+	// Write tombstone entry.
+	_, err := p.writeLogEntry(AppendSeriesEntry(nil, SeriesEntryTombstoneFlag, id, nil))
+	if err != nil {
+		return err
+	}
+
+	// Mark tombstone in memory.
+	p.index.Delete(id)
+
+	return nil
+}
+
+// IsDeleted returns true if the ID has been deleted before.
+func (p *SeriesPartition) IsDeleted(id uint64) bool {
```

Comment:
Do series that have never been added to this partition return `true`? Does this cover series that have been deleted and also series that have never existed?